src/semidiscretization/semidiscretization_coupled.jl

```diff
@@ -0,0 +1,257 @@
+
+    summary_footer(io)
+    println(io, "\n")
+  end
+end
+
+function Base.summary(semi::SemidiscretizationCoupled)
+  _, _, solver, _ = mesh_equations_solver_cache(semi.semis[1])
+  summary(solver)
+end
+
+
+@inline Base.ndims(semi::SemidiscretizationCoupled) = ndims(semi.semis[1])
+
+@inline nsystems(semi::SemidiscretizationCoupled) = length(semi.semis)
+
+@inline Base.real(semi::SemidiscretizationCoupled) = promote_type(real.(semi.semis)...)
+
+@inline Base.eltype(semi::SemidiscretizationCoupled) = promote_type(eltype.(semi.semis)...)
+
+@inline function ndofs(semi::SemidiscretizationCoupled)
+  sum(ndofs, semi.semis)
+end
+
+@inline function nelements(semi::SemidiscretizationCoupled)
+  return sum(semi.semis) do semi_
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semi_)
+
+    nelements(mesh, solver, cache)
+  end
+end
+
+@inline function total_volume(semi::SemidiscretizationCoupled)
+  sum(semi.semis) do semi_
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semi_)
+    total_volume(mesh, solver, cache)
+  end
+end
+
+
+function compute_coefficients(t, semi::SemidiscretizationCoupled)
+  @unpack u_indices = semi
+
+  u_ode = Vector{real(semi)}(undef, u_indices[end][end])
+
+  for i in 1:nsystems(semi)
+    # Call `compute_coefficients` in `src/semidiscretization/semidiscretization.jl`
+    u_ode[u_indices[i]] .= compute_coefficients(t, semi.semis[i])
+  end
+
+  return u_ode
+end
+
+
+function allocate_coupled_boundary_conditions(semi::AbstractSemidiscretization)
+  n_boundaries = 2 * ndims(semi)
+  mesh, equations, solver, _ = mesh_equations_solver_cache(semi)
+
+  for direction in 1:n_boundaries
+    boundary_condition = semi.boundary_conditions[direction]
+
+    allocate_coupled_boundary_condition(boundary_condition, direction, mesh, equations, solver)
+  end
+end
+
+# Don't do anything for other BCs than BoundaryConditionCoupled
+function allocate_coupled_boundary_condition(boundary_condition, direction, mesh, equations, solver)
+end
+
+# In 2D
+function allocate_coupled_boundary_condition(boundary_condition::BoundaryConditionCoupled{2},
+                                             direction, mesh, equations, dg::DGSEM)
+  if direction in (1, 2)
+    cell_size = size(mesh, 2)
+  else
+    cell_size = size(mesh, 1)
+  end
+
+  boundary_condition.u_boundary = Array{Float64, 3}(undef, nvariables(equations), nnodes(dg),
+                                                    cell_size)
+end
+
+
+function rhs!(du_ode, u_ode, semi::SemidiscretizationCoupled, t)
+  @unpack u_indices = semi
+
+  time_start = time_ns()
+
+  @trixi_timeit timer() "copy to coupled boundaries" begin
+    for semi_ in semi.semis
+      copy_to_coupled_boundary(semi_.boundary_conditions, u_ode, semi)
+    end
+  end
+
+  # Call rhs! for each semidiscretization
+  for i in 1:nsystems(semi)
+    u_loc  = @view u_ode[u_indices[i]]
+    du_loc = @view du_ode[u_indices[i]]
+
+    rhs!(du_loc, u_loc, semi.semis[i], t)
+  end
```

Comment:
This is not type stable - keyword lispy tuple programming

Reply:
This got changed in PR 1505. Once that is merged with this PR it should be resolved.

Reply:
@ranocha OK to resolve?

Reply:
You don't since you use the same semidiscretization (same type) twice. You will get it when you get to more complicated setups. Please add it to a TODO list for now.

Reply:
Tracked in #1520.